src/backend-c/runtime/cast_queue.c

```diff
@@ -0,0 +1,62 @@
+#include "assert.h"
+#include "cast_queue.h"
+#include "runtime.h"
+
+
+cast_queue* allocate_cast_queue(){
+  cast_queue *cq = GC_NEW(cast_queue);
+  cq->capacity = 0;
+  cq->front = 0;
+  cq->rear = 0;
+  return cq;
+}
+
+void cast_queue_enqueue(cast_queue* cq, grift_obj addr, grift_obj ty){
+  assert(cq->rear <= cq->capacity);
+  assert(cq->rear >= cq->front);
+
+  if (cq->rear == cq->capacity) {
+    if (cq->rear == 0) {
+      // Lazy allocation
+      cast_queue_resize(cq, CAST_QUEUE_INIT_CAPACITY);
+    } else {
+      cast_queue_resize(cq, 2 * cq->rear);
+    }
+  }
```

Comment:
Good point on getting rid of lazy allocation. I did some benchmarking and it seems that I can allocate a 10k queue without noticeable change in the runtime. I will set the initial capacity to 1k and allocate right away. 

Regarding the different strategy comment, this queue is already meant to be global. LIFO is different from my semantics and I am not sure how much it affects it so I will not consider it for now. Circular FIFO sounds like a good idea but it might be overkill in this case. I hypothesize that the queue should not see a frequent enqueues after dequeues.

Reply:
I think if the queue is "global" (ie used for every cast in the program), then you have a space leak. Correct? Someone could write a program that performs an unbounded number of enqueues by writing a loop that allocates an array, does some work the causes a whole bunch of enqueues followed by an equal number of dequeues. Then allocates a new array that is imprecise and does it all again in the next iteration of the loop. 

I think you have to reuse the space in the queue somehow. I was imagining that this code would have to deallocate the queue after each cast to not leak. If you have a global queue, the queue itself must reclaim used space in some way. A stack base implementation of a LIFO queue and a circular FIFO queue both solve this problem... If it is indeed a problem.

Reply:
The space is being reused on resize as discussed offline. I changed the queue to be circular anyway.